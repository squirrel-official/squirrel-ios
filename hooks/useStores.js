/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { createContext, useContext } from 'react';

import RootStore from '../stores/RootStore';

export const storesContext = createContext({
	rootStore: new RootStore()
});

export const useStores = () => useContext(storesContext);
