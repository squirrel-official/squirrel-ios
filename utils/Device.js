/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import compareVersions from 'compare-versions';
import Constants from 'expo-constants';
import * as Device from 'expo-device';
import { Platform } from 'react-native';

import iOSProfile from './profiles/ios';
import iOS10Profile from './profiles/ios10';
import iOS12Profile from './profiles/ios12';
import iOSFmp4Profile from './profiles/iosFmp4';

export function getAppName() {
	return `Jellyfin Mobile (${Device.osName})`;
}

export function getSafeDeviceName() {
	const safeName = Constants.deviceName
	// Replace non-ascii apostrophe with single quote (default on iOS)
		.replace(/â€™/g, '\'')
	// Remove all other non-ascii characters
		.replace(/[^\x20-\x7E]/g, '')
	// Trim whitespace
		.trim();
	if (safeName) {
		return safeName;
	}

	return Device.modelName || 'Jellyfin Mobile Device';
}

export function getDeviceProfile({ enableFmp4 = false } = {}) {
	if (Platform.OS === 'ios') {
		if (compareVersions.compare(Platform.Version, '11', '<')) {
			return iOS10Profile;
		} else if (compareVersions.compare(Platform.Version, '13', '<')) {
			return iOS12Profile;
		} else if (enableFmp4) {
			return iOSFmp4Profile;
		} else {
			return iOSProfile;
		}
	}
	// TODO: Add Android support
	return {};
}

export function isCompact({ height = 500 } = {}) {
	return height < 480;
}

// Does the platform support system level themes: https://docs.expo.io/versions/latest/sdk/appearance/
export function isSystemThemeSupported() {
	return (Platform.OS === 'ios' && compareVersions.compare(Platform.Version, '12', '>')) ||
		(Platform.OS === 'android' && compareVersions.compare(Platform.Version, '9', '>'));
}
