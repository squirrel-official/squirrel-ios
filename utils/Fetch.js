/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

export function fetchWithTimeout(url, timeout) {
	const abortController = new AbortController();
	const { signal } = abortController;

	const timeoutId = setTimeout(() => {
		console.log('fetch timed out, aborting');
		abortController.abort();
	}, timeout);

	return fetch(url, { signal })
		.finally(() => {
			clearTimeout(timeoutId);
		});
}
