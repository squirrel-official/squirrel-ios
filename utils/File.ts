/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import * as FileSystem from 'expo-file-system';

/**
 * Checks if a path exists, and creates a directory (including missing intermediate directories) if it does not.
 * @param path A uri of a local directory
 */
export async function ensurePathExists(path: string) {
	const info = await FileSystem.getInfoAsync(path);
	if (!info.exists) {
		await FileSystem.makeDirectoryAsync(path, { intermediates: true });
	}
}
