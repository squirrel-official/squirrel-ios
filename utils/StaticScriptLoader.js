/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { Asset } from 'expo-asset';
import { readAsStringAsync } from 'expo-file-system';

const loadStaticFile = async (asset) => {
	const [{ localUri }] = await Asset.loadAsync(asset);
	return readAsStringAsync(localUri);
};

class Loader {
	scripts = {
		NativeAudioPlayer: '',
		NativeVideoPlayer: '',
		NativeShell: '',
		ExpoRouterShim: ''
	}

	async load() {
		// Load NativeShell plugins
		this.scripts.NativeAudioPlayer = await loadStaticFile(require('../assets/js/plugins/NativeAudioPlayer.staticjs'));
		this.scripts.NativeVideoPlayer = await loadStaticFile(require('../assets/js/plugins/NativeVideoPlayer.staticjs'));
		// Load the NativeShell
		this.scripts.NativeShell = await loadStaticFile(require('../assets/js/NativeShell.staticjs'));
		// Load the RouterShim
		this.scripts.ExpoRouterShim = await loadStaticFile(require('../assets/js/ExpoRouterShim.staticjs'));

		return this.scripts;
	}
}

const StaticScriptLoader = new Loader();

export default StaticScriptLoader;
