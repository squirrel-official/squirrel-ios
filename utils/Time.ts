/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
const TICKS_PER_MS = 10000;

export const ticksToMs = (ticks = 0) => ticks / TICKS_PER_MS;

export const msToTicks = (ms = 0) => ms * TICKS_PER_MS;
