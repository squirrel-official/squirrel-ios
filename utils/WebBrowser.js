/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import * as WebBrowser from 'expo-web-browser';

import Colors from '../constants/Colors';

export async function openBrowser(url, options) {
	const finalOptions = Object.assign({
		toolbarColor: Colors.blackish,
		controlsColor: Colors.blue
	}, options);

	try {
		await WebBrowser.openBrowserAsync(url, finalOptions);
	} catch (err) {
		// Workaround issue where swiping browser closed does not dismiss it.
		// https://github.com/expo/expo/issues/6918
		if (err.message === 'Another WebBrowser is already being presented.') {
			try {
				await WebBrowser.dismissBrowser();
				return WebBrowser.openBrowserAsync(url, finalOptions);
			} catch (retryErr) {
				console.warn('Could not dismiss and reopen browser', retryErr);
			}
		} else {
			console.warn('Could not open browser', err);
		}
	}
}
