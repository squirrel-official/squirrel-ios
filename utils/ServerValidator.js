/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import normalizeUrl from 'normalize-url';

import { fetchWithTimeout } from './Fetch';

const TIMEOUT_DURATION = 5000; // timeout request after 5s

export const parseUrl = (host = '', port = '') => {
	if (!host?.trim()) {
		throw new Error('host cannot be blank');
	}

	// Normalize the entered url
	host = normalizeUrl(host, { stripWWW: false });

	// Parse the host as a url
	const url = new URL(host);

	// Override the port if provided
	if (port) {
		url.port = port;
	}
	return url;
};

export const fetchServerInfo = async (server = {}) => {
	const serverUrl = server.urlString || getServerUrl(server);
	const infoUrl = `${serverUrl}system/info/public`;
	console.log('info url', infoUrl);

	const responseJson = await fetchWithTimeout(infoUrl, TIMEOUT_DURATION)
		.then(response => {
			if (!response.ok) {
				throw new Error(`Error response status [${response.status}] received from ${infoUrl}`);
			}
			return response.json();
		});
	console.log('response', responseJson);

	return responseJson;
};

export const getServerUrl = (server = {}) => {
	if (!server?.url?.href) {
		throw new Error('Cannot get server url for invalid server', server);
	}

	// Strip the query string or hash if present
	let serverUrl = server.url.href;
	if (server.url.search || server.url.hash) {
		const endUrl = server.url.search || server.url.hash;
		serverUrl = serverUrl.substring(0, serverUrl.indexOf(endUrl));
	}

	// Ensure the url ends with /
	if (!serverUrl.endsWith('/')) {
		serverUrl += '/';
	}

	console.log('getServerUrl:', serverUrl);
	return serverUrl;
};

export const validateServer = async (server = {}) => {
	try {
		// Does the server have a valid url?
		getServerUrl(server);
	} catch (err) {
		return {
			isValid: false,
			message: 'invalid'
		};
	}

	try {
		const responseJson = await fetchServerInfo(server);
		const isValid = responseJson.ProductName === 'squirrel Server';
		const answer = {
			isValid
		};
		if (!isValid) {
			answer.message = 'invalidProduct';
		}
		return answer;
	} catch (err) {
		return {
			isValid: false,
			message: 'noConnection'
		};
	}
};
