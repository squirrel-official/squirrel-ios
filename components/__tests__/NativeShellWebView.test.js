/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import { render } from '@testing-library/react-native';
import React from 'react';

import { useStores } from '../../hooks/useStores';
import { getAppName, getDeviceProfile, getSafeDeviceName } from '../../utils/Device';
import NativeShellWebView from '../NativeShellWebView';

jest.mock('../../hooks/useStores');
useStores.mockImplementation(() => ({
	rootStore: {
		serverStore: {
			servers: [{
				info: {
					Version: '10.8.0'
				}
			}]
		},
		settingStore: {
			activeServer: 0
		}
	}
}));

jest.mock('../../utils/Device');
getAppName.mockImplementation(() => 'Jellyfin Mobile');
getDeviceProfile.mockImplementation(() => ({}));
getSafeDeviceName.mockImplementation(() => 'Test Device');

describe('NativeShellWebView', () => {
	it('should render correctly', () => {
		const { toJSON } = render(
			<NativeShellWebView />
		);

		expect(toJSON()).toMatchSnapshot();
	});
});
