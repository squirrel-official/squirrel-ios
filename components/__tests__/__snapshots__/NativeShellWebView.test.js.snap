// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NativeShellWebView should render correctly 1`] = `
<RCTScrollView
  collapsable={false}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onLayout={[Function]}
  refreshControl={
    <RefreshControlMock
      enabled={true}
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  showsHorizontalScrollIndicator={false}
  showsVerticalScrollIndicator={false}
  style={
    Object {
      "flex": 1,
      "height": "100%",
    }
  }
>
  <RCTRefreshControl />
  <View>
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCWebView
        allowsInlineMediaPlayback={true}
        cacheEnabled={true}
        decelerationRate={0.998}
        injectedJavaScript={null}
        injectedJavaScriptBeforeContentLoaded="
window.ExpoAppInfo = {
	appName: 'squirrel Mobile',
	appVersion: 'mock',
	deviceId: 'undefined',
	deviceName: 'Test Device'
};

window.ExpoAppSettings = {
	isPluginSupported: true,
	isNativeVideoPlayerEnabled: undefined,
	isExperimentalNativeAudioPlayerEnabled: undefined,
	isExperimentalDownloadsEnabled: undefined
};

window.ExpoVideoProfile = {};

function postExpoEvent(event, data) {
	window.ReactNativeWebView.postMessage(JSON.stringify({
		event: event,
		data: data
	}));
}








window.onerror = console.error;

true;
"
        injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
        injectedJavaScriptForMainFrameOnly={true}
        javaScriptEnabled={true}
        mediaPlaybackRequiresUserAction={false}
        messagingEnabled={true}
        onContentProcessDidTerminate={[Function]}
        onHttpError={[Function]}
        onLoadingError={[Function]}
        onLoadingFinish={[Function]}
        onLoadingProgress={[Function]}
        onLoadingStart={[Function]}
        onMessage={[Function]}
        onScroll={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        showsHorizontalScrollIndicator={false}
        showsVerticalScrollIndicator={false}
        source={
          Object {
            "uri": undefined,
          }
        }
        style={
          Array [
            Object {
              "flex": 1,
              "overflow": "hidden",
            },
            Object {
              "backgroundColor": "#ffffff",
            },
            Object {
              "flex": 1,
              "height": 1334,
            },
          ]
        }
        textInteractionEnabled={true}
        useSharedProcessPool={true}
        useWebKit={true}
      />
    </View>
  </View>
</RCTScrollView>
`;
